lib/backend/test/atomicwrite.go

```diff
@@ -0,0 +1,688 @@
+				}
+
+				n++
+
+				_, err = bk.AtomicWrite(ctx, backend.ConditionalAction{
+					Key:       counterKey,
+					Condition: backend.Revision(item.Revision),
+					Action: backend.Put(backend.Item{
+						Value: []byte(strconv.Itoa(n)),
+					}),
+				})
+
+				if err != nil {
+					if errors.Is(err, backend.ErrConditionFailed) {
+						continue
+					}
+
+					// should never happen unless test is malformed or backend is offline
+					panic(fmt.Sprintf("unexpected error writing counter: %v", err))
+				}
+
+				localIncrements++
+			}
+
+			if localIncrements < increments/workers {
+				// should never happen unless test is malformed or backend is offline
+				panic(fmt.Sprintf("worker halted with %d/%d local increments (this is a bug)", localIncrements, increments/workers))
+			}
+		}()
+	}
+
+	wg.Wait()
+
+	counterItem, err := bk.Get(ctx, counterKey)
+	require.NoError(t, err)
+
+	n, err := strconv.Atoi(string(counterItem.Value))
+	require.NoError(t, err)
+	require.Equal(t, increments, n)
+}
+
+// testAtomicWriteNonConflicting verifies that non-conflicting but overlapping transactions all succeed
+// on the first attempt when running concurrently, meaning that backends that treat overlap as conflict (e.g. dynamo)
+// handle such conflicts internally.
+func testAtomicWriteNonConflicting(t *testing.T, newBackend AtomicWriteConstructor) {
+	const workers = 60
+
+	bk, _, err := newBackend()
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	results := make(chan error, workers)
+
+	commonKey := []byte("/nc-common")
+
+	itemKey := func(i int) []byte {
+		return []byte(fmt.Sprintf("/nc-item-%d", i))
+	}
+
+	_, err = bk.Put(ctx, backend.Item{
+		Key:   commonKey,
+		Value: []byte("c1"),
+	})
+	require.NoError(t, err)
+
+	for i := 0; i < workers; i++ {
+		go func(i int) {
+			_, err := bk.AtomicWrite(ctx, []backend.ConditionalAction{
+				{
+					Key:       commonKey,
+					Condition: backend.Exists(),
+					Action:    backend.Nop(),
+				},
+				{
+					Key:       itemKey(i),
+					Condition: backend.Whatever(),
+					Action: backend.Put(backend.Item{
+						Value: []byte("v1"),
+					}),
+				},
+			}...)
+
+			results <- err
+		}(i)
+	}
+
+	timeout := time.After(time.Minute)
+
+	for i := 0; i < workers; i++ {
+		select {
+		case err := <-results:
+			require.NoError(t, err, trace.DebugReport(err))
+		case <-timeout:
+			require.FailNow(t, "timeout waiting for workers to finish", "iteration=%d", i)
+		}
+	}
+
+	for i := 0; i < workers; i++ {
+		item, err := bk.Get(ctx, itemKey(i))
+		require.NoError(t, err)
+		require.Equal(t, []byte("v1"), item.Value)
+	}
+}
+
+// testAtomicWriteOther verifies some minor edge-cases that may not be covered by other tests. Specifically,
+// it verifies that Item.Key has no effect on writes or subsequent reads, and that ineffectual writes still
+// update the value of revision.
+func testAtomicWriteOther(t *testing.T, newBackend AtomicWriteConstructor) {
```

Comment:
Should we test (or explicitly not support) an AtomicWrite that only has Nop actions (for atomic checks across multiple keys)?